cohere.ts

```diff
@@ -31,23 +33,68 @@ class Cohere implements CohereService {
   private makeRequest(
     model: string,
     endpoint: string,
-    data: models.cohereParameters,
+    data: models.cohereParameters
   ): Promise<models.cohereResponse<models.responseBody>> {
     return API.post(`/${model}${endpoint}`, data);
   }
 
-  public generate(model: string, config: models.generate): Promise<models.cohereResponse<models.text>> {
-    return this.makeRequest(model, ENDPOINT.GENERATE, config) as Promise<models.cohereResponse<models.text>>;
+  public generate(
+    model: string,
+    config: models.generate
+  ): Promise<models.cohereResponse<models.text>> {
+    return this.makeRequest(model, ENDPOINT.GENERATE, config) as Promise<
+      models.cohereResponse<models.text>
+    >;
   }
 
-  public embed(model: string, config: models.embed): Promise<models.cohereResponse<models.embeddings>> {
-    return this.makeRequest(model, ENDPOINT.EMBED, config) as Promise<models.cohereResponse<models.embeddings>>;
-  }
+  public embed(
+    model: string,
+    config: models.embed
+  ): Promise<models.cohereResponse<models.embeddings>> {
+    const createBatches = (array: string[]) => {
+      const result = [];
+      for (const value of array) {
+        const lastArray = result[result.length - 1];
+        if (!lastArray || lastArray.length === COHERE_EMBED_BATCH_SIZE) {
+          result.push([value]);
+        } else {
+          lastArray.push(value);
+        }
+      }
+      return result;
+    };
 
-  public chooseBest(model: string, config: models.chooseBest): Promise<models.cohereResponse<models.scores>> {
-    return this.makeRequest(model, ENDPOINT.CHOOSE_BEST, config) as Promise<models.cohereResponse<models.scores>>;
+    return Promise.all(
+      createBatches(config.texts).map(
+        (texts) =>
+          this.makeRequest(model, ENDPOINT.EMBED, {
+            ...config,
+            texts,
+          }) as Promise<models.cohereResponse<models.embeddings>>
+      )
+    ).then((results) => {
+      let embeddings: number[][] = [];
+      results.forEach((result) => {
+        embeddings = embeddings.concat(result.body.embeddings);
+      });
+
+      const response: models.cohereResponse<models.embeddings> = {
+        statusCode: results[0].statusCode,
```

Comment:
out of curiosity do you happen to know the conventions for batch requests like if 1/4 fail, do we just return the happiest status code + the incomplete results?

Reply:
in this case, `Promise.all` will throw error if any batch fails and I am not keen on any convention for batch requests, but I think if one fails, we should treat the whole request as `failed`. The user is not aware that behind the scenes it's a batch but rather one request. Returning incomplete results might be confusing and not what the user expects